ntoskrnl/mm/ARM3/mdlsup.c

```diff
@@ -74,6 +74,10 @@ MiMapLockedPagesInUserSpace(
         DPRINT1("FIXME: Need to check for large pages\n");
     }
 
+    Status = PsChargeProcessNonPagedPoolQuota(Process, sizeof(MMVAD_LONG));
+    if (!NT_SUCCESS(Status))
+        goto ErrorWithoutQuotaReturn;
```

Comment:
We could avoid that new goto label altogether by doing something like that:

```c
    Vad = ExAllocatePoolWithTag(NonPagedPool, sizeof(MMVAD_LONG), 'ldaV');
    if (Vad == NULL)
    {
        Status = STATUS_INSUFFICIENT_RESOURCES;
        goto Error;
    }

    Status = PsChargeProcessNonPagedPoolQuota(Process, sizeof(MMVAD_LONG));
    if (!NT_SUCCESS(Status))
    {
        ExFreePoolWithTag(Vad, 'ldaV');
        Vad = NULL;
        goto Error;
    }

// Code continues as usual.... and then, in the error cleanup at the end:

Error:
    if (AddressSpaceLocked)
    {
        MmUnlockAddressSpace(&Process->Vm);
    }
    if (Vad != NULL)
    {
        PsReturnProcessNonPagedPoolQuota(Process, sizeof(MMVAD_LONG));
        ExFreePoolWithTag(Vad, 'ldaV');
    }
    ExRaiseStatus(Status);
```

Note that in the PsChargeProcessNonPagedPoolQuota failure path, we free the Vad and reset the pointer to NULL so that, in the cleanup path, we don't again release quota and double-free the Vad.
This ensures a *logical* ordering of steps (one first allocates memory before doing any quota increase).
In any case, the idea here was that at cleanup time, both the Vad-freeing + the quota return have to happen at the same time.

The other modifications in your PR may be adapted with that idea.